File: src/lib/escrow.ts
import {
  Connection,
  PublicKey,
  Transaction,
  SystemProgram,
  LAMPORTS_PER_SOL,
  TransactionInstruction,
  Keypair,
} from '@solana/web3.js';
import * as borsh from 'borsh';
import bs58 from 'bs58';
import { connection } from '@/lib/solana';

/**
 * SECURITY: Convert SOL to lamports with proper precision handling
 * JavaScript floating point can cause issues like 0.1 * 1e9 = 99999999.99999999
 * This function rounds to 9 decimal places first to avoid precision loss
 */
export function solToLamports(amountSOL: number): number {
  // Round to 9 decimal places (maximum precision for SOL)
  const roundedSOL = Math.round(amountSOL * 1e9) / 1e9;
  // Convert to lamports using integer math after rounding
  return Math.round(roundedSOL * LAMPORTS_PER_SOL);
}

/**
 * Convert lamports to SOL with precision handling
 */
export function lamportsToSOL(lamports: number): number {
  return lamports / LAMPORTS_PER_SOL;
}

// Escrow Program ID (deployed on Devnet)
// For now using a placeholder - replace with actual deployed program ID
// Uses lazy initialization to avoid errors at import time
let _escrowProgramId: PublicKey | null = null;

export function getEscrowProgramId(): PublicKey {
  if (!_escrowProgramId) {
    const programId = process.env.NEXT_PUBLIC_ESCROW_PROGRAM_ID;
    if (!programId) {
      // Use System Program as fallback (won't work for actual escrow, but won't crash)
      _escrowProgramId = SystemProgram.programId;
    } else {
      _escrowProgramId = new PublicKey(programId);
    }
  }
  return _escrowProgramId;
}

// Legacy export - use getEscrowProgramId() instead
export const ESCROW_PROGRAM_ID = {
  toBuffer: () => getEscrowProgramId().toBuffer(),
  toBase58: () => getEscrowProgramId().toBase58(),
  equals: (other: PublicKey) => getEscrowProgramId().equals(other),
} as unknown as PublicKey;

// Escrow status enum
export enum EscrowStatus {
  Funding = 0,
  Active = 1,
  Completed = 2,
  Cancelled = 3,
  Disputed = 4,
}

// Escrow account data structure
export interface EscrowAccount {
  isInitialized: boolean;
  proposalId: string;
  agentWallet: PublicKey;
  fundingGoal: number;
  totalFunded: number;
  milestoneCount: number;
  milestonesCompleted: number;
  status: EscrowStatus;
  authority: PublicKey;
  bump: number;
}

// Instruction types
export enum EscrowInstruction {
  InitializeEscrow = 0,
  Fund = 1,
  CompleteMilestone = 2,
  CancelProposal = 3,
  ClaimRefund = 4,
}

// Find PDA for escrow account
export async function findEscrowPDA(proposalId: string): Promise<[PublicKey, number]> {
  return PublicKey.findProgramAddressSync(
    [Buffer.from('escrow'), Buffer.from(proposalId)],
    getEscrowProgramId()
  );
}

// Find PDA for funding record
export async function findFundingRecordPDA(
  escrow: PublicKey,
  backer: PublicKey
): Promise<[PublicKey, number]> {
  return PublicKey.findProgramAddressSync(
    [Buffer.from('funding'), escrow.toBuffer(), backer.toBuffer()],
    getEscrowProgramId()
  );
}

// Create initialize escrow instruction
export function createInitializeEscrowInstruction(
  agent: PublicKey,
  escrowPDA: PublicKey,
  proposalId: string,
  fundingGoal: number,
  milestoneCount: number
): TransactionInstruction {
  // Encode instruction data
  const data = Buffer.alloc(1 + 32 + 8 + 1);
  data.writeUInt8(EscrowInstruction.InitializeEscrow, 0);

  // Write proposal ID (padded to 32 bytes)
  const proposalIdBuffer = Buffer.alloc(32);
  proposalIdBuffer.write(proposalId);
  proposalIdBuffer.copy(data, 1);

  // Write funding goal in lamports
  data.writeBigUInt64LE(BigInt(solToLamports(fundingGoal)), 33);

  // Write milestone count
  data.writeUInt8(milestoneCount, 41);

  return new TransactionInstruction({
    keys: [
      { pubkey: agent, isSigner: true, isWritable: true },
      { pubkey: escrowPDA, isSigner: false, isWritable: true },
      { pubkey: SystemProgram.programId, isSigner: false, isWritable: false },
    ],
    programId: getEscrowProgramId(),
    data,
  });
}

// Create fund instruction
export function createFundInstruction(
  backer: PublicKey,
  escrowPDA: PublicKey,
  fundingRecordPDA: PublicKey,
  amount: number
): TransactionInstruction {
  const data = Buffer.alloc(1 + 8);
  data.writeUInt8(EscrowInstruction.Fund, 0);
  data.writeBigUInt64LE(BigInt(solToLamports(amount)), 1);

  return new TransactionInstruction({
    keys: [
      { pubkey: backer, isSigner: true, isWritable: true },
      { pubkey: escrowPDA, isSigner: false, isWritable: true },
      { pubkey: fundingRecordPDA, isSigner: false, isWritable: true },
      { pubkey: SystemProgram.programId, isSigner: false, isWritable: false },
    ],
    programId: getEscrowProgramId(),
    data,
  });
}

// Create complete milestone instruction
export function createCompleteMilestoneInstruction(
  authority: PublicKey,
  escrowPDA: PublicKey,
  agentWallet: PublicKey,
  milestoneIndex: number
): TransactionInstruction {
  const data = Buffer.alloc(2);
  data.writeUInt8(EscrowInstruction.CompleteMilestone, 0);
  data.writeUInt8(milestoneIndex, 1);

  return new TransactionInstruction({
    keys: [
      { pubkey: authority, isSigner: true, isWritable: false },
      { pubkey: escrowPDA, isSigner: false, isWritable: true },
      { pubkey: agentWallet, isSigner: false, isWritable: true },
    ],
    programId: getEscrowProgramId(),
    data,
  });
}

// Create cancel proposal instruction
export function createCancelProposalInstruction(
  authority: PublicKey,
  escrowPDA: PublicKey
): TransactionInstruction {
  const data = Buffer.alloc(1);
  data.writeUInt8(EscrowInstruction.CancelProposal, 0);

  return new TransactionInstruction({
    keys: [
      { pubkey: authority, isSigner: true, isWritable: false },
      { pubkey: escrowPDA, isSigner: false, isWritable: true },
    ],
    programId: getEscrowProgramId(),
    data,
  });
}

// Create claim refund instruction
export function createClaimRefundInstruction(
  backer: PublicKey,
  escrowPDA: PublicKey,
  fundingRecordPDA: PublicKey
): TransactionInstruction {
  const data = Buffer.alloc(1);
  data.writeUInt8(EscrowInstruction.ClaimRefund, 0);

  return new TransactionInstruction({
    keys: [
      { pubkey: backer, isSigner: true, isWritable: true },
      { pubkey: escrowPDA, isSigner: false, isWritable: true },
      { pubkey: fundingRecordPDA, isSigner: false, isWritable: true },
    ],
    programId: getEscrowProgramId(),
    data,
  });
}

// Helper to check if escrow program is deployed
export async function isEscrowProgramDeployed(connection: Connection): Promise<boolean> {
  try {
    const accountInfo = await connection.getAccountInfo(getEscrowProgramId());
    return accountInfo !== null && accountInfo.executable;
  } catch {
    return false;
  }
}

// For MVP: Simple escrow using direct transfer (no program needed)
// This is a temporary solution until the program is deployed
export async function createSimpleFundingTransaction(
  connection: Connection,
  fromWallet: PublicKey,
  toWallet: PublicKey,
  amountSOL: number
): Promise<Transaction> {
  const transaction = new Transaction();

  transaction.add(
    SystemProgram.transfer({
      fromPubkey: fromWallet,
      toPubkey: toWallet,
      lamports: solToLamports(amountSOL),
    })
  );

  const { blockhash } = await connection.getLatestBlockhash();
  transaction.recentBlockhash = blockhash;
  transaction.feePayer = fromWallet;

  return transaction;
}

// Escrow mode configuration
let _platformAuthority: PublicKey | null = null;

function getPlatformAuthority(): PublicKey {
  if (!_platformAuthority) {
    const authority = process.env.NEXT_PUBLIC_PLATFORM_AUTHORITY;
    if (!authority) {
      _platformAuthority = SystemProgram.programId;
    } else {
      _platformAuthority = new PublicKey(authority);
    }
  }
  return _platformAuthority;
}

export const ESCROW_MODE = {
  // Set to true when escrow program is deployed
  USE_PROGRAM: false,

  // Platform authority for milestone verification (getter function)
  get PLATFORM_AUTHORITY(): PublicKey {
    return getPlatformAuthority();
  },
};

// Platform escrow wallet - funds are held here until milestone completion
// Uses lazy initialization to avoid errors when env vars aren't set
let _platformEscrowWallet: PublicKey | null = null;

export function getPlatformEscrowWallet(): PublicKey {
  if (!_platformEscrowWallet) {
    const walletAddress = process.env.NEXT_PUBLIC_ESCROW_WALLET || process.env.NEXT_PUBLIC_PLATFORM_AUTHORITY;
    if (!walletAddress || walletAddress === '11111111111111111111111111111111') {
      throw new Error('NEXT_PUBLIC_ESCROW_WALLET environment variable not configured');
    }
    _platformEscrowWallet = new PublicKey(walletAddress);
  }
  return _platformEscrowWallet;
}

// Legacy export for backwards compatibility - only use when escrow is configured
export const PLATFORM_ESCROW_WALLET = {
  toBase58: () => {
    const wallet = process.env.NEXT_PUBLIC_ESCROW_WALLET;
    if (!wallet || wallet === '11111111111111111111111111111111') {
      throw new Error('Escrow wallet not configured');
    }
    return wallet;
  }
};

// Check if escrow is properly configured
export function isEscrowConfigured(): boolean {
  const escrowWallet = process.env.NEXT_PUBLIC_ESCROW_WALLET;
  return !!escrowWallet && escrowWallet !== '11111111111111111111111111111111';
}

// ============================================================================
// Server-side escrow functions
// ============================================================================

/**
 * Load escrow keypair from ESCROW_WALLET_PRIVATE_KEY environment variable
 * The private key should be base58 encoded
 * @returns Keypair for the escrow wallet
 * @throws Error if ESCROW_WALLET_PRIVATE_KEY is not set
 */
export function getEscrowKeypair(): Keypair {
  const privateKeyBase58 = process.env.ESCROW_WALLET_PRIVATE_KEY;
  if (!privateKeyBase58) {
    throw new Error('ESCROW_WALLET_PRIVATE_KEY environment variable is not set');
  }
  const privateKeyBytes = bs58.decode(privateKeyBase58);
  return Keypair.fromSecretKey(privateKeyBytes);
}

/**
 * Transfer funds from escrow wallet to agent wallet
 * @param agentWallet - The agent's wallet address (base58 string)
 * @param amountSOL - Amount to transfer in SOL
 * @returns Transaction signature
 */
export async function releaseToAgent(agentWallet: string, amountSOL: number): Promise<string> {
  const escrowKeypair = getEscrowKeypair();
  const agentPubkey = new PublicKey(agentWallet);

  const transaction = new Transaction();
  transaction.add(
    SystemProgram.transfer({
      fromPubkey: escrowKeypair.publicKey,
      toPubkey: agentPubkey,
      lamports: solToLamports(amountSOL),
    })
  );

  const { blockhash } = await connection.getLatestBlockhash();
  transaction.recentBlockhash = blockhash;
  transaction.feePayer = escrowKeypair.publicKey;

  transaction.sign(escrowKeypair);

  const signature = await connection.sendRawTransaction(transaction.serialize());
  // SECURITY: Use 'finalized' commitment for permanent transaction confirmation
  await connection.confirmTransaction(signature, 'finalized');

  return signature;
}

/**
 * Transfer funds from escrow wallet back to backer wallet (refund)
 * @param backerWallet - The backer's wallet address (base58 string)
 * @param amountSOL - Amount to refund in SOL
 * @returns Transaction signature
 */
export async function refundToBacker(backerWallet: string, amountSOL: number): Promise<string> {
  const escrowKeypair = getEscrowKeypair();
  const backerPubkey = new PublicKey(backerWallet);

  const transaction = new Transaction();
  transaction.add(
    SystemProgram.transfer({
      fromPubkey: escrowKeypair.publicKey,
      toPubkey: backerPubkey,
      lamports: solToLamports(amountSOL),
    })
  );

  const { blockhash } = await connection.getLatestBlockhash();
  transaction.recentBlockhash = blockhash;
  transaction.feePayer = escrowKeypair.publicKey;

  transaction.sign(escrowKeypair);

  const signature = await connection.sendRawTransaction(transaction.serialize());
  // SECURITY: Use 'finalized' commitment for permanent transaction confirmation
  await connection.confirmTransaction(signature, 'finalized');

  return signature;
}

/**
 * Verify a transaction was successful on-chain
 * @param signature - Transaction signature to verify
 * @param expectedRecipient - Expected recipient wallet address
 * @param expectedAmountSOL - Expected amount in SOL
 * @returns Verification result
 */
export async function verifyTransactionSuccess(
  signature: string,
  expectedRecipient: string,
  expectedAmountSOL: number
): Promise<{ valid: boolean; error?: string }> {
  try {
    const tx = await connection.getTransaction(signature, {
      commitment: 'finalized',
      maxSupportedTransactionVersion: 0,
    });

    if (!tx) {
      return { valid: false, error: 'Transaction not found' };
    }

    if (tx.meta?.err) {
      return { valid: false, error: 'Transaction failed on-chain' };
    }

    // Verify recipient received funds
    const accountKeys = tx.transaction.message.getAccountKeys();
    let recipientIndex = -1;
    for (let i = 0; i < accountKeys.length; i++) {
      if (accountKeys.get(i)?.toBase58() === expectedRecipient) {
        recipientIndex = i;
        break;
      }
    }

    if (recipientIndex === -1) {
      return { valid: false, error: 'Recipient not found in transaction' };
    }

    // Check amount received
    const preBalance = tx.meta?.preBalances?.[recipientIndex] || 0;
    const postBalance = tx.meta?.postBalances?.[recipientIndex] || 0;
    const receivedLamports = postBalance - preBalance;
    const receivedSOL = receivedLamports / LAMPORTS_PER_SOL;

    // Allow 0.1% tolerance for rounding
    const tolerance = expectedAmountSOL * 0.001;
    if (Math.abs(receivedSOL - expectedAmountSOL) > tolerance) {
      return { valid: false, error: `Amount mismatch: expected ${expectedAmountSOL}, got ${receivedSOL}` };
    }

    return { valid: true };
  } catch (error) {
    console.error('[Escrow] Transaction verification error:', error);
    return { valid: false, error: 'Verification failed' };
  }
}

/**
 * Get the current balance of the escrow wallet
 * @returns Balance in SOL
 */
export async function getEscrowBalance(): Promise<number> {
  const escrowKeypair = getEscrowKeypair();
  const balance = await connection.getBalance(escrowKeypair.publicKey);
  return balance / LAMPORTS_PER_SOL;
}

File: contracts/escrow.rs
// AgentFund Escrow Smart Contract
// This Solana program handles milestone-based escrow for AI agent proposals
//
// Key Features:
// - Create escrow accounts for proposals
// - Accept funding from multiple backers
// - Release funds when milestones are verified
// - Refund if proposal fails/cancelled
//
// To deploy: Use Anchor framework or native Solana program

use solana_program::{
    account_info::{next_account_info, AccountInfo},
    entrypoint,
    entrypoint::ProgramResult,
    msg,
    program_error::ProgramError,
    pubkey::Pubkey,
    program::{invoke, invoke_signed},
    system_instruction,
    sysvar::{rent::Rent, Sysvar},
};
use borsh::{BorshDeserialize, BorshSerialize};

// Program ID (replace with actual deployed address)
solana_program::declare_id!("AgentFund1111111111111111111111111111111111");

// Escrow account state
#[derive(BorshSerialize, BorshDeserialize, Debug)]
pub struct Escrow {
    pub is_initialized: bool,
    pub proposal_id: String,           // Off-chain proposal ID
    pub agent_wallet: Pubkey,          // Agent's wallet to receive funds
    pub funding_goal: u64,             // Target in lamports
    pub total_funded: u64,             // Current funded amount
    pub milestone_count: u8,           // Number of milestones (typically 4)
    pub milestones_completed: u8,      // How many milestones done
    pub status: EscrowStatus,          // Current status
    pub authority: Pubkey,             // Platform authority for disputes
    pub bump: u8,                      // PDA bump seed
}

#[derive(BorshSerialize, BorshDeserialize, Debug, PartialEq)]
pub enum EscrowStatus {
    Funding,      // Accepting funds
    Active,       // Funded, work in progress
    Completed,    // All milestones done
    Cancelled,    // Project cancelled, refunds available
    Disputed,     // Under dispute resolution
}

// Funding record for each backer
#[derive(BorshSerialize, BorshDeserialize, Debug)]
pub struct FundingRecord {
    pub backer: Pubkey,
    pub amount: u64,
    pub escrow: Pubkey,
    pub refunded: bool,
}

// Instructions
#[derive(BorshSerialize, BorshDeserialize, Debug)]
pub enum EscrowInstruction {
    /// Initialize a new escrow for a proposal
    /// Accounts:
    /// 0. [signer] Agent wallet
    /// 1. [writable] Escrow account (PDA)
    /// 2. [] System program
    InitializeEscrow {
        proposal_id: String,
        funding_goal: u64,
        milestone_count: u8,
    },

    /// Fund a proposal (send SOL to escrow)
    /// Accounts:
    /// 0. [signer] Backer wallet
    /// 1. [writable] Escrow account
    /// 2. [writable] Funding record account (PDA)
    /// 3. [] System program
    Fund {
        amount: u64,
    },

    /// Mark a milestone as complete and release funds
    /// Accounts:
    /// 0. [signer] Platform authority
    /// 1. [writable] Escrow account
    /// 2. [writable] Agent wallet (receives funds)
    CompleteMilestone {
        milestone_index: u8,
    },

    /// Cancel proposal and enable refunds
    /// Accounts:
    /// 0. [signer] Platform authority or agent
    /// 1. [writable] Escrow account
    CancelProposal,

    /// Claim refund (for cancelled proposals)
    /// Accounts:
    /// 0. [signer] Backer wallet
    /// 1. [writable] Escrow account
    /// 2. [writable] Funding record account
    ClaimRefund,
}

entrypoint!(process_instruction);

pub fn process_instruction(
    program_id: &Pubkey,
    accounts: &[AccountInfo],
    instruction_data: &[u8],
) -> ProgramResult {
    let instruction = EscrowInstruction::try_from_slice(instruction_data)?;

    match instruction {
        EscrowInstruction::InitializeEscrow { proposal_id, funding_goal, milestone_count } => {
            msg!("Instruction: InitializeEscrow");
            process_initialize_escrow(program_id, accounts, proposal_id, funding_goal, milestone_count)
        }
        EscrowInstruction::Fund { amount } => {
            msg!("Instruction: Fund");
            process_fund(program_id, accounts, amount)
        }
        EscrowInstruction::CompleteMilestone { milestone_index } => {
            msg!("Instruction: CompleteMilestone");
            process_complete_milestone(program_id, accounts, milestone_index)
        }
        EscrowInstruction::CancelProposal => {
            msg!("Instruction: CancelProposal");
            process_cancel_proposal(program_id, accounts)
        }
        EscrowInstruction::ClaimRefund => {
            msg!("Instruction: ClaimRefund");
            process_claim_refund(program_id, accounts)
        }
    }
}

fn process_initialize_escrow(
    program_id: &Pubkey,
    accounts: &[AccountInfo],
    proposal_id: String,
    funding_goal: u64,
    milestone_count: u8,
) -> ProgramResult {
    let account_info_iter = &mut accounts.iter();

    let agent_account = next_account_info(account_info_iter)?;
    let escrow_account = next_account_info(account_info_iter)?;
    let system_program = next_account_info(account_info_iter)?;

    if !agent_account.is_signer {
        return Err(ProgramError::MissingRequiredSignature);
    }

    // Derive PDA for escrow
    let (escrow_pda, bump) = Pubkey::find_program_address(
        &[b"escrow", proposal_id.as_bytes()],
        program_id,
    );

    if escrow_pda != *escrow_account.key {
        return Err(ProgramError::InvalidAccountData);
    }

    // Create escrow account
    let escrow = Escrow {
        is_initialized: true,
        proposal_id,
        agent_wallet: *agent_account.key,
        funding_goal,
        total_funded: 0,
        milestone_count,
        milestones_completed: 0,
        status: EscrowStatus::Funding,
        authority: *agent_account.key, // In production, use platform multisig
        bump,
    };

    escrow.serialize(&mut *escrow_account.data.borrow_mut())?;

    msg!("Escrow initialized for proposal");
    Ok(())
}

fn process_fund(
    program_id: &Pubkey,
    accounts: &[AccountInfo],
    amount: u64,
) -> ProgramResult {
    let account_info_iter = &mut accounts.iter();

    let backer_account = next_account_info(account_info_iter)?;
    let escrow_account = next_account_info(account_info_iter)?;
    let _funding_record_account = next_account_info(account_info_iter)?;
    let system_program = next_account_info(account_info_iter)?;

    if !backer_account.is_signer {
        return Err(ProgramError::MissingRequiredSignature);
    }

    let mut escrow = Escrow::try_from_slice(&escrow_account.data.borrow())?;

    if escrow.status != EscrowStatus::Funding {
        return Err(ProgramError::InvalidAccountData);
    }

    // Transfer SOL to escrow
    invoke(
        &system_instruction::transfer(
            backer_account.key,
            escrow_account.key,
            amount,
        ),
        &[backer_account.clone(), escrow_account.clone(), system_program.clone()],
    )?;

    escrow.total_funded += amount;

    // Check if fully funded
    if escrow.total_funded >= escrow.funding_goal {
        escrow.status = EscrowStatus::Active;
        msg!("Proposal fully funded!");
    }

    escrow.serialize(&mut *escrow_account.data.borrow_mut())?;

    msg!("Funded {} lamports", amount);
    Ok(())
}

fn process_complete_milestone(
    program_id: &Pubkey,
    accounts: &[AccountInfo],
    milestone_index: u8,
) -> ProgramResult {
    let account_info_iter = &mut accounts.iter();

    let authority_account = next_account_info(account_info_iter)?;
    let escrow_account = next_account_info(account_info_iter)?;
    let agent_account = next_account_info(account_info_iter)?;

    if !authority_account.is_signer {
        return Err(ProgramError::MissingRequiredSignature);
    }

    let mut escrow = Escrow::try_from_slice(&escrow_account.data.borrow())?;

    if escrow.status != EscrowStatus::Active {
        return Err(ProgramError::InvalidAccountData);
    }

    if *authority_account.key != escrow.authority {
        return Err(ProgramError::InvalidAccountData);
    }

    if milestone_index != escrow.milestones_completed {
        return Err(ProgramError::InvalidArgument);
    }

    // Calculate release amount (equal split per milestone)
    let release_amount = escrow.funding_goal / escrow.milestone_count as u64;

    // Transfer from escrow to agent
    **escrow_account.try_borrow_mut_lamports()? -= release_amount;
    **agent_account.try_borrow_mut_lamports()? += release_amount;

    escrow.milestones_completed += 1;

    if escrow.milestones_completed == escrow.milestone_count {
        escrow.status = EscrowStatus::Completed;
        msg!("All milestones completed!");
    }

    escrow.serialize(&mut *escrow_account.data.borrow_mut())?;

    msg!("Milestone {} completed, released {} lamports", milestone_index, release_amount);
    Ok(())
}

fn process_cancel_proposal(
    _program_id: &Pubkey,
    accounts: &[AccountInfo],
) -> ProgramResult {
    let account_info_iter = &mut accounts.iter();

    let authority_account = next_account_info(account_info_iter)?;
    let escrow_account = next_account_info(account_info_iter)?;

    if !authority_account.is_signer {
        return Err(ProgramError::MissingRequiredSignature);
    }

    let mut escrow = Escrow::try_from_slice(&escrow_account.data.borrow())?;

    // Only authority or agent can cancel
    if *authority_account.key != escrow.authority && *authority_account.key != escrow.agent_wallet {
        return Err(ProgramError::InvalidAccountData);
    }

    escrow.status = EscrowStatus::Cancelled;
    escrow.serialize(&mut *escrow_account.data.borrow_mut())?;

    msg!("Proposal cancelled");
    Ok(())
}

fn process_claim_refund(
    _program_id: &Pubkey,
    accounts: &[AccountInfo],
) -> ProgramResult {
    let account_info_iter = &mut accounts.iter();

    let backer_account = next_account_info(account_info_iter)?;
    let escrow_account = next_account_info(account_info_iter)?;
    let funding_record_account = next_account_info(account_info_iter)?;

    if !backer_account.is_signer {
        return Err(ProgramError::MissingRequiredSignature);
    }

    let escrow = Escrow::try_from_slice(&escrow_account.data.borrow())?;

    if escrow.status != EscrowStatus::Cancelled {
        return Err(ProgramError::InvalidAccountData);
    }

    let mut funding_record = FundingRecord::try_from_slice(&funding_record_account.data.borrow())?;

    if funding_record.backer != *backer_account.key {
        return Err(ProgramError::InvalidAccountData);
    }

    if funding_record.refunded {
        return Err(ProgramError::InvalidAccountData);
    }

    // Transfer refund
    **escrow_account.try_borrow_mut_lamports()? -= funding_record.amount;
    **backer_account.try_borrow_mut_lamports()? += funding_record.amount;

    funding_record.refunded = true;
    funding_record.serialize(&mut *funding_record_account.data.borrow_mut())?;

    msg!("Refunded {} lamports", funding_record.amount);
    Ok(())
}

// Tests
#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_escrow_status() {
        assert_eq!(EscrowStatus::Funding, EscrowStatus::Funding);
    }
}
